File: RecursiveDescentParsing.java
import java.util.*;

public class RecursiveDescentParsing {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        Dictionary<String, List<String>> grammar = new Hashtable<String, List<String>>();

        System.out.print("Enter the number of grammar: ");
        int n = scanner.nextInt();
        System.out.print("Enter an input string: ");
        String s = scanner.next();

        for (int i = 0; i < n; i++) {
            String key = scanner.next();
            String temp = scanner.next();
            String[] words = temp.split("|");
            List<String> list = new ArrayList<String>();
            for (String word : words) {
                list.add(word);
            }
            grammar.put(key, list);
        }

        parse(s, grammar);
    }

    private static boolean parse(String s, Dictionary<String, List<String>> grammar, String iterator, int index) {
        if (iterator == null) {
            iterator = "S";
        }

        return false;
    }
    private static boolean isNonTerminal(String symbol) {
        return symbol.length() == 1 && Character.isUpperCase(symbol.charAt(0));
    }
}

File: Recursive_descent_parsing.py
def backtrack(s, grammar, ite=[], index=0):
    if "".join(ite) == s: #check if iterator is equal to string
        return True
    
    if (len(ite) == 0): #check if iterator is empty
        return backtrack(s, grammar, [list(grammar.keys())[0]], index)
    
    if len(ite)<=index or len(s)<=index or len(ite)<=index: #check if index is out of range
        return False
    
    if ite[index] == s[index]: #check if matching
        flag = backtrack(s, grammar, ite, index+1)
        if flag:
            return True
    
    if ite[index] in grammar: #check if non-terminal
        for word in grammar[ite[index]]: #iterate through the words
            temp = ite.copy()
            temp.pop(index)
            for i in range(len(word)):
                temp.insert(index+i, word[i])
                
            #print("".join(temp)) #debugging
            flag = backtrack(s, grammar, temp, index)
            if flag:
                return True
            
    #check if mismatch
    return False

n = int(input("Enter the number of grammar: "))

grammar = {}

for i in range(n):
    key = input("Enter the key: ")
    words = input("Enter the words: ").split("|")
    for j in range(len(words)):
        words[j] = words[j].strip()
    grammar[key] = words
        
s = input("Enter the string: ")

start = input("Enter the start symbol: ")

if backtrack(s, grammar, [*start]):
    print("String is accepted")
else:
    print("String is not accepted")